### PR TITLE
Resolve shape via ops in broadcast_to operation

### PR DESCRIPTION
**Pull request by sampathweb**
Resolves the following issues -
1. When `ops.broadcast_to`, requires the shape to be resolved via `ops.shape` and fails when used with tensor.shape attribute for TensorFlow backend.
2. When multiplying `sample_weight` with `y_true`, it needs to be cast to the target tensor.  Many times, `self.dtype` is None and for regression metrics like `R2Score`, `y_true` is a float and `sample_weight` is an int, causing type mismatch in TensorFlow backend.
3. Remove the dtype specification in `self.num_samples` in `R2Score` metric initialization.  Having the dtype specified in the init, causes the variable to be put in CPU, causing CPU / GPU error when assigning value in `update_state` method.

There is still couple of more issues with how `sample_weight` is used in `iou_metric` that needs to be resolved.

### CODE REVIEW

File: keras_core/metrics/metric.py [hunk 1]
```diff
@@ -72,7 +72,9 @@ def update_state(self, y_true, y_pred, sample_weight=None):
             values = ops.cast(values, self.dtype)
             if sample_weight is not None:
                 sample_weight = ops.cast(sample_weight, self.dtype)
-                sample_weight = ops.broadcast_to(sample_weight, values.shape)
+                sample_weight = ops.broadcast_to(
+                    sample_weight, ops.shape(values)
+                )
```
Comment (chenmoneygithub) [inline keras_core/metrics/metric.py:77]: Actually maybe the explicit `broadcast_to` is not necessary, in the next line `ops.multiply` should automatically do the broadcasting. Not 100% sure tho.

Comment (sampathweb) [reply keras_core/metrics/metric.py:77]: This is in the docstring of the `metrics`, so it might be better to keep it explicit and might be useful in some cases.

Comment (chenmoneygithub) [reply keras_core/metrics/metric.py:77]: kk, sgtm!

**Comment by fchollet** [comment]
I'm going to merge this for the rerelease -- please add the unit test in a separate PR.

**Review by fchollet — APPROVED** [review]
